### PR TITLE
Update experiment setup script for free-forecast ensemble

### LINKED ISSUE #789: Update experiment setup script for free-forecast ensemble
**Description**
As we move towards the merger of GEFS into the GFS system, we need to be able to generate experiments with an arbitrary number of free-forecast ensemble members in addition to the number of EnKF members.

**Requirements**
Ability to generate experiments (using `setup_expt.py`) with a number of free-forecast ensemble members determined via a command-line switch similar to how the number of EnKF members is determined. The default should be 0, indicating no perturbed members (control only). This will require `config.base` to be updated to provide a corresponding new variable which will be used by all the job scripts later.

**Acceptance Criteria (Definition of Done)**
A `setup_expt.py` script that takes in a number of member for the free-forecast and sets an appropriate variable in `config.base`.

**Dependencies**
None

### PR DESCRIPTION
**Pull request by XianwuXue-NOAA**
**Description**

As we move towards the merger of GEFS into the GFS system, we need to be able to generate experiments using setup_expt.py with an arbitrary number of free-forecast ensemble members in addition to the number of EnKF members.

**Type of change**

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update (add "gefs" to setup_expt.py)

**How Has This Been Tested?**

Tested on Hera for ATM only GEFS forecast-only

  
**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes

### CODE REVIEW

File: ush/rocoto/setup_expt.py [hunk 1]
```diff
@@ -30,7 +30,8 @@ def fill_COMROT(host, inputs):
 
     fill_modes = {
         'cycled': fill_COMROT_cycled,
-        'forecast-only': fill_COMROT_forecasts
+        'forecast-only': fill_COMROT_forecasts,
+        'gefs': fill_COMROT_gefs
```
Comment (WalterKolczynski-NOAA) [inline ush/rocoto/setup_expt.py]: Instead of adding a new mode, there should be a new command line option to specify the number of ensemble members (with a default of 0). While we are starting with forecast-only, at some point in the future we will want to run full cycling and also have a forecast ensemble.

File: ush/rocoto/setup_expt.py [hunk 2]
```diff
@@ -231,18 +241,21 @@ def input_args():
                           choices=['warm', 'cold'], required=False, default='cold')
 
     # cycled mode additional arguments
-    cycled.add_argument('--resens', help='resolution of the ensemble model forecast',
-                        type=int, required=False, default=192)
-    cycled.add_argument('--nens', help='number of ensemble members',
-                        type=int, required=False, default=20)
+    for subp in [cycled, gefs]:
```
Comment (WalterKolczynski-NOAA) [inline ush/rocoto/setup_expt.py]: New options need to be created for the forecast ensemble rather than overloading the options that were created for EnKF. Existing options should be renamed `--nens-enkf` and `--res-enkf` to avoid confusion.

File: ush/rocoto/setup_expt.py [hunk 3]
```diff
@@ -147,12 +155,12 @@ def edit_baseconfig(host, inputs):
         "@APP@": inputs.app,
     }
 
-    if inputs.mode in ['cycled']:
+    if inputs.mode in ['cycled', 'gefs']:
         extend_dict = {
             "@CASEENS@": f'C{inputs.resens}',
```
Comment (WalterKolczynski-NOAA) [inline ush/rocoto/setup_expt.py]: This variable name will need to be changed in config.base and everywhere downstream to something like `$CASE_ENKF` to avoid confusion with new resolution variable for GEFS.

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]
This needs to be reworked to be more inline with what we discussed previously (unless I'm forgetting or missed a conversation).

Also, the PR title and description need a *lot* of work, as they are practically non-existent right now.

File: ush/rocoto/setup_expt.py (hunk 1 above)
Comment (WalterKolczynski-NOAA) [reply ush/rocoto/setup_expt.py]: Actually, we might want to use -1 as the default (meaning high-res only). For now assume GEFS-resolution only if the number of members >= 0, though we'll probably need a new switch in the future to turn off the high-res instead, since we'll want the option to everything (GFS, GEFS, GDAS, EnKF) together at some point, but we don't always want to run high-res when we run GEFS.

File: ush/rocoto/setup_expt.py (hunk 2 above)
Comment (XianwuXue-NOAA) [reply ush/rocoto/setup_expt.py]: OK. I will add --nens-gefs, --res-gefs to setup_expt.py and also make some changes in config.gefs.

File: ush/rocoto/setup_expt.py (hunk 3 above)
Comment (XianwuXue-NOAA) [reply ush/rocoto/setup_expt.py]: How about I change to, however, I keep inputs.resens and inputs.nens, then users don't need to change command options

    if inputs.mode in ['cycled']:
        extend_dict = {
            "@CASEENS@": f'C{inputs.resens}',
            "@NMEM_ENKF@": inputs.nens,
        }
    elif inputs.mode in ['gefs']:
        extend_dict = {
            "@CASEENS_GEFS@": f'C{inputs.resens}',
            "@NMEM_GEFS@": inputs.nens,
        }
    elif inputs.mode in ['forecast-only', 'gefs']:
        extend_dict = {
            "@DO_AERO@": inputs.aerosols,
        }

File: ush/rocoto/setup_expt.py (hunk 2 above)
Comment (XianwuXue-NOAA) [reply ush/rocoto/setup_expt.py]: How about the changes below:

    if inputs.mode in ['cycled']:
        extend_dict = {
            "@CASEENS@": f'C{inputs.resens}',
            "@NMEM_ENKF@": inputs.nens,
        }
    elif inputs.mode in ['gefs']:
        extend_dict = {
            "@CASEENS_GEFS@": f'C{inputs.resens}',
            "@NMEM_GEFS@": inputs.nens,
        }
    elif inputs.mode in ['forecast-only', 'gefs']:
        extend_dict = {
            "@DO_AERO@": inputs.aerosols,
        }

File: ush/rocoto/setup_expt.py (hunk 3 above)
Comment (WalterKolczynski-NOAA) [reply ush/rocoto/setup_expt.py]: This suggested code won't even work properly, and I still want to update the options so it is clear they belong to EnKF. As always when we change a script signature, we will make an announcement when the change is merged so everyone is aware.

Comment (WalterKolczynski-NOAA) [reply ush/rocoto/setup_expt.py]: Plus, I still don't want to add gefs as a new mode.

Comment (XianwuXue-NOAA) [reply ush/rocoto/setup_expt.py]: If you don't want to add gefs as a new mode, do you have any suggestions?

> This suggested code won't even work properly, and I still want to update the options so it is clear they belong to EnKF. As always when we change a script signature, we will make an announcement when the change is merged so everyone is aware.

Yes, I should also change the configure file. If you agree these changes, I will work on it.

Comment (XianwuXue-NOAA) [reply ush/rocoto/setup_expt.py]: > Plus, I still don't want to add gefs as a new mode.

I think adding gefs as a new mode is clear. Any suggestions?

Comment (WalterKolczynski-NOAA) [reply ush/rocoto/setup_expt.py]: Yes, it's in the first change request (along with why)

Comment (XianwuXue-NOAA) [reply ush/rocoto/setup_expt.py]: Originally, I want to reuse the existing variables. However, based on this conversation, I changed my mind to add new variables. If I misunderstood, please let me know

**Comment by XianwuXue-NOAA** [comment]
@WalterKolczynski-NOAA I have modified the variables and also added new options for GEFS.

File: .gitignore [hunk 4]
```diff
@@ -122,7 +122,7 @@ sorc/wafs_gcip.fd
 sorc/wafs_grib2_0p25.fd
 sorc/wafs_makewafs.fd
 sorc/wafs_setmissing.fd
-
+sorc/upp.fd
```
Comment (aerorahul) [inline .gitignore]: how is this related to this PR?

File: parm/config/config.gefs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#!/bin/ksh -x
```
Comment (aerorahul) [inline parm/config/config.gefs:1]: Which job is sourcing this file?
It may be needed in subsequent PR's.
Consider adding when it is used.

File: ush/rocoto/workflow_utils.py [hunk 6]
```diff
@@ -272,6 +272,9 @@ def source_configs(configs, tasks):
         elif 'wave' in task:
             files.append(find_config(f'config.wave', configs))
             files.append(find_config(f'config.{task}', configs))
+        elif 'gefs' in task:
+            files.append(find_config(f'config.gefs', configs))
+            files.append(find_config(f'config.{task}', configs))
```
Comment (aerorahul) [inline ush/rocoto/workflow_utils.py]: There are no gefs tasks currently, so this is a placeholder.
Since there is work ongoing on refactoring setup_workflow, this could be held till actual tasks are being added.

File: parm/config/config.base.emc.dyn [hunk 7]
```diff
@@ -224,6 +225,8 @@ if [[ "$CDUMP" == "gdas" ]] ; then
    export FTSFS=10
 elif [[ "$CDUMP" == "gfs" ]] ; then
    export FHCYC=24
+elif [[ "$CDUMP" == "gefs" ]] ; then
+   export FHCYC=24
```
Comment (aerorahul) [inline parm/config/config.base.emc.dyn]: May be replace lines 223-230 with `case ${CDUMP} in` statements and condense `gfs` and `gefs`

File: parm/config/config.base.emc.dyn [hunk 8]
```diff
@@ -207,7 +207,8 @@ export OPS_RES="C768" # Do not change
 # Resolution specific parameters
 export LEVS=128
 export CASE="@CASECTL@"
-export CASE_ENKF="@CASEENS@"
+export CASE_ENKF="@CASEENS_ENKF@"
+export CASE_GEFS="@CASEENS_GEFS@"
```
Comment (aerorahul) [inline parm/config/config.base.emc.dyn]: I am still not convinced we will have a use case for running GEFS that is a resolution different than the cycled ensemble resolution.

**Review by aerorahul — CHANGES_REQUESTED** [review]
please see comments.

File: .gitignore (hunk 4 above)
Comment (XianwuXue-NOAA) [reply .gitignore]: No relate to this PR, so I will delete it.

File: parm/config/config.gefs (hunk 5 above)
Comment (XianwuXue-NOAA) [reply parm/config/config.gefs:1]: Currently, this config file is just for placeholder. If there is no need for now until we need this file, I can delete this file.

File: ush/rocoto/workflow_utils.py (hunk 6 above)
Comment (XianwuXue-NOAA) [reply ush/rocoto/workflow_utils.py]: Currently, I use the existing getic, init and fcst tasks for GEFS, later, I will add recenter task for GEFS. OK, I can delete these lines until actual tasks are being added

File: parm/config/config.base.emc.dyn (hunk 7 above)
Comment (XianwuXue-NOAA) [reply parm/config/config.base.emc.dyn]: OK. I will do it.

File: parm/config/config.base.emc.dyn (hunk 8 above)
Comment (XianwuXue-NOAA) [reply parm/config/config.base.emc.dyn]: @aerorahul Any suggestions? Should I change it back to 'export CASE_ENKF="@CASEENS@"' and the corresponding changes for the options or keep this changes?
